docs/cli/ilab-model-serve.md

```diff
@@ -0,0 +1,101 @@
+# Design for `ilab model serve`
+
+## Background
+
+With the [request from the community](https://github.com/instructlab/instructlab/issues/1106) for `ilab` to serve different backends such as [vllm](https://docs.vllm.ai/en/stable/) and the [cli redesign](https://github.com/instructlab/dev-docs/pull/48), this design doc's purpose is to flesh out the behavior of the `ilab model serve` command.
+
+Specifically the design of the command, its flags/options, and the how a new serving backend will be tested within the project.
+
+## Design
+
+Since the subject of the `ilab model serve` command is a model, regardless of the format of the model, every command takes in the `--model` flag or uses its default value in the config.
+
+`ilab model serve` has a subcommand for each backend it can serve models with. As of this draft, the two backends `ilab` would serve with are `llama-cpp` and `vllm`.
+
+This would lead to the commands:
+
+- `ilab model serve llama-cpp`
+- `ilab model serve vllm`
+
+The flags for `ilab model serve` are either backend agnostic or specific to a subcommand and are prefixed to indicate that. The intention behind this choice is to ensure there is no mix up of flags between backends.
+
+```console
+ilab model serve
+|
+|_______ --model
+|
+|_______ (other backend agnostic flags)
+|
+|_______ llama-cpp
+|        |
+|        |_______ --llama-cpp-num-gpus
+|        |
+|        |_______ --llama-cpp-max-ctx-size
+|        |
+|        |_______ (other llama-cpp specific args)
+|
+|_______ vllm
+|        |
+|        |_______ --vllm-chat-template
+|        |
+|        |_______ (other popular vllm-specific specific args)
+|        |
+|        |_______ --vllm-args
+```
+
+## Command Examples
+
+### Bare-bones but model specific command
+
+```shell
+ilab model serve --model <PATH> 
+```
+
+- Serves the model at `<PATH>`.
+- If the `<PATH>` is the path for a `.gguf` file then `llama-cpp` is automatically used as the model serving backend.
+- If the path is a directory, `vllm` is automatically used as the backend.
+- The path provided to `vllm` can either be to a Hugging Face repo or a directory with `.safetensors`.
+
+### Bare-bones command
+
+```shell
+ilab model serve
+```
+
+- This command has the same behavior as the one above but the `--model-path` is whatever the default model path is in the config. This is the existing behavior of `ilab serve` today.
+
+### Llama-cpp backend specific commands
+
+```shell
+ilab model serve llama-cpp --model <PATH> --llama-cpp-gpu-layers 4 --llama-cpp-max-ctx-size 1024
```

Comment:
A point I've raised before, I'm not sure if it makes sense to prefix everything with `--llama-cpp` if we know we are using the llama-cpp backend.

Reply:
The reason I made it so specific is that I'm scarred by the flags for training that exist in the codebase as of 6/7/24. It's not clear which training flags belong with the mac implementation of training vs the linux implementation and on top of that the flags are very similar.

An example of how the flags could be mixed up. Current ilab serve with llama-cpp has a flag for `max-ctx-size`. `vllm` has a flag for model context length called `--max-model-len` (https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html#command-line-arguments-for-the-server). 

We could do the following possibilities:
1. Have a flag for `--max-model-len` and `--max-ctx-size` similar to how we have it in current `ilab train`. I think this option is hardest on the user.

2. Have one flag for context size called `--max-ctx-size` whose value is filtered down to each backend. An issue then becomes what do you do with flags that one backend has that one doesn't. Also you've now added this flag into the set of flags that both backends can use. 

In my current design I propose we make backend agnostic flags those flags have nothing to do with the configuration of the backend itself and more for the configuration of `ilab` and what the backends are operating on (like the model path).

3. Prefix all flags related to a backend with the backend name. I think this is the most clear for the users with the downside that it's more verbose.

@cdoern what do you think about how the flags for training might look. Training will have a similar set of possible subcommands with numerous options.

Reply:
> In my current design I propose we make backend agnostic flags those flags have nothing to do with the configuration of the backend itself and more for the configuration of ilab and what the backends are operating on (like the model path).
 +1 this is the right approach to create and use generic flags yet still have a mechanism for provider specific flags.

Reply:
I'm open to building generic flags, but it feels like we're creating a third layer of abstraction, which might introduce some ambiguity. Users will have to juggle raw flags from llama-cpp, as well as flags from ilab, which could complicate their mental model when writing commands. However, I do like the idea of re-using sensible flags and handling the backend mapping internally.

In the end, we could adopt two strategies:

* Backend-Agnostic Flags: Select the most common flags that make sense across multiple backends.
* Backend-Specific Flags: Use --vllm-args (or similar) for backend-specific configurations.

There might still be issues, such as someone passing ilab model serve --backend llama-cpp --llama-cpp-max-ctx-size 1 --llama-cpp-args "llama-cpp-max-ctx-size 2". However, we can address this through proper validation and error messaging.

Reply:
After doing some more research I'm removing the backend specific prefixes from the backend specific flags. The way we're using `click` for subcommands, flags between backends cannot be shared.

So running `ilab model serve vllm --help` would yield a different message than `ilab model serve --help` and `ilab model serve llama-cpp --help`. 

For this reason I'm not worried about users mixing up arguments as long as there are subcommands.

This is another reason why I believe going subcommands instead of having a  `--backend` flag to choose the backend.